app/assets/src/components/views/amr_heatmap/AMRHeatmapControls.jsx

```diff
@@ -35,12 +47,44 @@ export default class AMRHeatmapControls extends React.Component {
     return controlsList;
   }
 
+  renderLegend() {
+    // Don't render a color legend if the heatmap is still loading
+    if (!this.props.isDataReady) {
+      return;
+    }
+    // This code is here so that when we need the legend to change
+    // because the scale has changed, React will unload the current
+    // legend from the DOM, and render the new, proper one when state updates.
+    if (
+      this.props.maxValueForLegend !== this.state.legend.max ||
+      this.props.selectedOptions.scale !== this.state.legend.scale
+    ) {
+      this.setState({
```

Comment:
Normally, you would put this kind of code (which updates the state based on the props) inside `componentDidUpdate` or `getDerivedStateFromProps`. For a short discussion on the difference, see: https://twitter.com/dan_abramov/status/977181473424932864.

It's risky to put `setState` calls inside the `render()` function, because `setState` will trigger a re-render, which if you're not careful will call setState again, causing an infinite loop. Changing the state inside the render function is generally considered bad practice in React, where render functions are supposed to be "pure" and not cause side effects.

However, in this case, it looks like you can just directly use `this.props` inside `<SequentialLegendVis>`, without storing it in the state at all.

Reply:
I moved this code to `componentDidUpdate()`. I did want to trigger a rerender to dismiss the `SequentialLegendVis`, then another render to put the new one in. (I couldn't get `SequentialLegendVis` to update the legend based on just props, I was getting some ugly visual artifacts). But yeah, I totally get wanting to keep render functions pure.

Reply:
If it's not too much work, can you modify `SequentialLegendVis` to update on new props?

It looks like `SequentialLegendVis` only implements `componentDidMount` but not `componentDidUpdate`. If you add a `updateOptions` method to `SequentialLegend` that lets you update `this.options` with new options, and then in SequentialLegendVis do:

```
componentDidUpdate() {
  this.legend.updateOptions({
    scale: this.props.scale,
    ...
  })
  this.legend.update()
}
```

that might be enough to get it to work.

Reply:
I'll see if I can get that to work!

Reply:
Got it!